### ISSUE TITLE
Search- Search shows User A submits to User B when User B submits to User A in the same report

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#44093]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.86-0
**Reproducible in staging?:** Y
**Reproducible in production?:** Y 
**If this was caught during regression testing, add the test name, ID and link from TestRail:** https://expensify.testrail.io/index.php?/tests/view/4647855
**Issue reported by:** Applause - Internal Team

## Action Performed:
1. Go to staging.new.expensify.com
2. [User A] Submit an expense to User B
3. [User B] Submit an expense to User A
4. [User A or User B] Go to Search > Shared
5. Look for the expense submitted by User B

## Expected Result:
From and To field for the expense submitted by User B will show User B in From field and User A in To field

## Actual Result:
From and To field for the expense submitted by User B will show User A in From field and User B in To field, when it is submitted from User B and User A
The Total field also does not make sense either. The amount sent by User A should be indicated by a negative or better indication so that the total amount correctly reflects the sum of both expenses

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [ ] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

</details> 

https://github.com/Expensify/App/assets/78819774/7c662746-72a2-417c-8070-078915168a3f

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by lanitochka17** [comment]
@johncschuster  FYI I haven't added the External label as I wasn't 100% sure about this issue. Please take a look and add the label if you agree it's a bug and can be handled by external contributors

**Comment by c3024** [comment]
## Proposal
### Please re-state the problem that we are trying to solve in this issue.
Search shows incorrect same `from` and `to` for all expenses between a set of users though some of the expenses have `from` and to the other way.
### What is the root cause of that problem?
If the sum of all the IOUs between two users A and B comes to be "User A owes X", all the transactions have same `from` 'B' and `to` 'A' for all transactions passed here
https://github.com/Expensify/App/blob/b021195b864211efbaaefe188226b2830b91b7d8/src/components/SelectionList/Search/ReportListItem.tsx#L188
so `from` 'B' and `to` 'A' is same for all transactions.
### What changes do you think we should make in order to solve the problem?
We should swap the `from` and `to` based on the `amount` in the `transaction`.
![Screenshot 2024-06-22 at 11 07 32 AM](https://github.com/Expensify/App/assets/102477862/fb7e7efc-5a47-4ab8-b5a1-79603c86eaf6)
![Screenshot 2024-06-22 at 11 07 14 AM](https://github.com/Expensify/App/assets/102477862/d01c618b-6baa-4c67-aeca-19741b3d874f)
something like this
```js
const participantFrom = item.amount > 0 ? item.from : item.to;
    const participantTo = item.amount > 0 ? item.to : item.from;
    const participantFromDisplayName = item.amount > 0 ? item.formattedFrom : item.formattedTo;
    const participantToDisplayName = item.amount > 0 ? item.formattedTo : item.formattedFrom;
```
in `TransactionListItemRow` and pass these here 
https://github.com/Expensify/App/blob/b021195b864211efbaaefe188226b2830b91b7d8/src/components/SelectionList/Search/TransactionListItemRow.tsx#L313-L321
and here
https://github.com/Expensify/App/blob/b021195b864211efbaaefe188226b2830b91b7d8/src/components/SelectionList/Search/TransactionListItemRow.tsx#L228-L232
### What alternative solutions did you explore? (Optional)

**Comment by brunovjk** [comment]
I will review the proposal first tomorrow

**Comment by brunovjk** [comment]
Thanks for the [proposal](https://github.com/Expensify/App/issues/44093#issuecomment-2183821669) @c3024. Sorry but, your root cause is not very clear to me:
> If the sum of all the IOUs between two users A and B comes to be "User A owes X", all the transactions have the same from 'B' and to 'A'

I can reproduce and see this, but why does it happen? Can we fix this?

**Comment by c3024** [comment]
That is how the backend sends the response for Search page @brunovjk

**Comment by brunovjk** [comment]
> That is how the backend sends the response for Search page @brunovjk

I see, @c3024 can/should we fix it there?

---------------------------

**1.** I tested your solution and works:
<details>
<summary>screenshot_1</summary>


![image](https://github.com/Expensify/App/assets/95647348/594a86d8-51da-4984-a8b1-f70d4ed39d8c)

</details>

Sorry, but it seems like a workaround, don't you think?

---------------------------
 
**2.** When we reduce the screen, the `FROM` and `TO` columns are replaced by a title:
<details>
<summary>screenshot_2</summary>



![image](https://github.com/Expensify/App/assets/95647348/ce28802d-796f-4f06-b78a-54bcde00b850)


</details>

This title also changes depending on "who owes less". Do you think we should update this as well? Maybe always show the name of the logged in user first, and in the total amount indicate "who owes who", as stated in the issue description.

---------------------------
 
**3.** Lastly, in the issue description it says:
_''The Total field also does not make sense either. The amount sent by User A should be indicated by a negative or better indication so that the total amount correctly reflects the sum of both expenses"_
Have you thought anything about this?

Thanks.

**Comment by c3024** [comment]
>can/should we fix it there?

I think a backend fix needs to be coupled with some frontend changes as well because for the narrow layout we are just taking the first transactions `from` and `to` as the user who submitted the more expense value and the other user respectively.

https://github.com/Expensify/App/blob/67f92c6fbe5dce7aff7ddd82b196bbc306ab9bf8/src/components/SelectionList/Search/ReportListItem.tsx#L89-L90

If we were to send the transactions `from` and `to` differently for different transactions between a set of users the first transaction need not be the user who submitted the more expense value and this needs to be checked.

I don't think the solution I suggested is a workaround. 😆 

>When we reduce the screen, the FROM and TO columns are replaced by a title:

I think this is consistent with individual rows in the wider screen so in my opinion this is fine. I don't think negative values would look good here. You might want to tag the design team if you think swapping the user positions would be better.


>The Total field also does not make sense either. The amount sent by User A should be indicated by a negative or better indication so that the total amount correctly reflects the sum of both expenses

I think this is clear enough. The Total field is the net amount (with the one who submitted more expense value or who owes less) on the left or `from` side. Once we correctly show the `from` and `to` as suggested in my fix this confusion should go away.

You can make a comment with the `ribbon eyes ribbon` without approving a proposal to get an internal engineer assigned to this issue and they can provide their input or you can start a discussion on Slack if you need more input.

**Comment by brunovjk** [comment]
> I think a backend fix needs to be coupled with some frontend changes as well because for the narrow layout we are just taking the first transactions `from` and `to` as the user who submitted the more expense value and the other user respectively.
> 
> https://github.com/Expensify/App/blob/67f92c6fbe5dce7aff7ddd82b196bbc306ab9bf8/src/components/SelectionList/Search/ReportListItem.tsx#L89-L90
> 
> If we were to send the transactions `from` and `to` differently for different transactions between a set of users the first transaction need not be the user who submitted the more expense value and this needs to be checked.

I think this would be a most complete solution to the issue, correct? We would attack the root cause.

However, in any case, your initial solution is simpler and also solves the first point of the issue: _`From` and `To` field for the expense submitted by User B will show User B in `From` field and User A in `To` field_.

Now about the second point:
> The Total field also does not make sense either. The amount sent by User A should be indicated by a negative or better indication so that the total amount correctly reflects the sum of both expenses

Also think this is clear enough, let's confirm.

🎀👀🎀

**Comment by brunovjk** [comment]
@cristipaval please take a look at my [last comment](https://github.com/Expensify/App/issues/44093#issuecomment-2189713944). I tagged you to confirm a few things before approving the [proposal](https://github.com/Expensify/App/issues/44093#issuecomment-2183821669).

-------------

We need to fix:
1. _"`From` and `To` field for the expense submitted by User B will show User A in `From` field and User B in `To` field, when it is submitted from User B and User A"_
2. _"The Total field also does not make sense either. The amount sent by User A should be indicated by a negative or better indication so that the total amount correctly reflects the sum of both expenses"_

For the first point @c3024 's [proposal](https://github.com/Expensify/App/issues/44093#issuecomment-2183821669) works well, still, I wonder if a backend/frontend solution might better resolve the root cause.
Could you please confirm the expected outcome for the second point above?

Thank you!

**Comment by cristipaval** [comment]
@luacmartins I think you have more knowledge around **Search** functionality. Could you have a look at this one or maybe take it over? 👀

**Comment by luacmartins** [comment]
I agree that we need to implement some backend fixes for this. I can work on that.

**Comment by c3024** [comment]
Will you change the individual transactions' `from` and `to` to actual expense submitter and submitted respectively with `amount` always positive in a `transaction`?

**Comment by luacmartins** [comment]
I'm flipping the `managerID` and `accountID` if the report is an IOU report and the amount is negative. I wasn't gonna change the amount since that seems to be displayed correctly to the user

**Comment by luacmartins** [comment]
Maybe @Expensify/design could chime in on this one. Should we show the amount as negative to the user if the request on an IOU report is not theirs?
![Screenshot 2024-06-26 at 12 35 50 PM](https://github.com/Expensify/App/assets/22219519/f9d7db07-fb99-4797-babb-6981f535514c)

**Comment by c3024** [comment]
There is a field `formattedAmount` which has the non-negative value of amount which is used to show the `total` in a row.

But the `transaction` has another field called `amount`.

Presently all transactions between a pair of users have the same `from` (the user who submitted more expense value or who is owed a net amount in other words) and `to`  but this `amount` field is negative if the transaction is in the other way.

I think if we swap the from and to of individual transactions it would be correct to change the individual transaction amount sign as well though it will not affect what's displayed to the end user because the total is based on the formattedAmount.

**Comment by shawnborton** [comment]
I don't think it makes sense to use a negative here personally. I might be missing something but what is actually wrong with how this is implemented? Isn't this following the exact logic/patterns that we have for IOUs with multiple expenses across both parties?

**Comment by shawnborton** [comment]
I think one issue is that we aren't using the correct report title here. Why is this called "IOU" in the grouped view and now "User A owes User B" or something like that?

**Comment by shawnborton** [comment]
Just tested with Carlos, and yeah, the IOU report actually has a title, not just "IOU": 
![CleanShot 2024-06-26 at 21 04 35@2x](https://github.com/Expensify/App/assets/2319350/8d9eb5dd-3d82-406b-ac3f-6ad38aa4531b)

**Comment by luacmartins** [comment]
Also for [this](https://github.com/Expensify/App/issues/44093#issuecomment-2189713944), we'll start returning the `managerID` and `accountID` in the report object so that we can use that instead of the first transaction's values.

**Comment by dannymcclain** [comment]
Agree with everything Shawn said above. Nothing seems wrong with this except the title of the group of expenses.

**Comment by luacmartins** [comment]
Ok, PR to return the flipped accountIDs and add `managerID` and `accountID` to the report object is in review. Once that's done, we'll need to update reportListItem to use the `managerID` and `accountID` from the report and not the first transaction

**Comment by luacmartins** [comment]
@brunovjk API PR is deployed to prod, we can continue work to use the report data instead of first transaction data

**Comment by c3024** [comment]
Updated the proposal for handling the remaining required frontend fix for `ExpenseItemHeaderNarrow` @brunovjk

**Comment by brunovjk** [comment]
Reviewing 🔁

**Comment by brunovjk** [comment]
@luacmartins @c3024's [proposal](https://github.com/Expensify/App/issues/44093#issuecomment-2183821669) looks good to me.

🎀👀🎀 C+ reviewed

**Comment by brunovjk** [comment]
> Issue not reproducible during KI retests. (First week)

The main issue here was resolved by BE, so it is no longer reproducible. We are making some necessary changes to FE to keep up, currently testing PR..

**Comment by johncschuster** [comment]
Acknowledged! I will pay this out after the regression threshold is crossed 👍 

@brunovjk / @c3024, can you complete the BZ Checklist before then? Thank you!

**Comment by brunovjk** [comment]
@luacmartins should we wait to fix [#45013](https://github.com/Expensify/App/issues/45013) to proceed with payment/closing here? Thanks.

**Comment by luacmartins** [comment]
Sure, let's do that

**Comment by brunovjk** [comment]
@luacmartins  would you mind clarifying [this](https://github.com/Expensify/App/issues/44093#issuecomment-2228619320)? Thank you.

**Comment by luacmartins** [comment]
@johncschuster correct, we're waiting for [this PR](https://github.com/Expensify/App/pull/45105) to hit production and start the countdown on the 7 day regression period. We can update the payment date once that PR is deployed to prod. I also left a comment on [this issue](https://github.com/Expensify/App/issues/45013#issuecomment-2228931662) so we don't double pay.

**Comment by johncschuster** [comment]
Awesome. Thanks for clarifying, @luacmartins! I'll be standing by for payment 👍

**Comment by johncschuster** [comment]
Looks like the linked PR just hit prod. Payment should be issued on July 24 assuming no regressions.

**Comment by johncschuster** [comment]
No regressions so far. We're still on track to pay this out on the 24th 👍

**Comment by brunovjk** [comment]
* [x]  [@brunovjk] The PR that introduced the bug has been identified. Link to the PR: **N/A**
* [x]  [@brunovjk] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment: **N/A**
* [x]  [@brunovjk] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion: **N/A**
* [x]  [@brunovjk] Determine if we should create a regression test for this bug. **Yes**
* [x]  [@brunovjk] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again. 

**Regression Test Proposal**
- Launch New Expensify app.
- Create a new workspace.
- [User A] Submit an expense for amount 'X' (e.g., 10 units) to [User B].
- [User B] Submit an expense for amount 'Y' (e.g., 20 units) to [User A].
- [User A] Submit an expense to the workspace.
- Go to Search > Shared.
- Verify that:
  - The From and To fields are correctly displayed for each expense (e.g., User B in From field and User A in To field for the expense from User B).
  - The total amount accurately reflects the sum of both user expenses with correct indications.
  - The workspace expense is grouped correctly and shows that it was submitted to the workspace.
 
Do we agree 👍 or 👎

**Comment by johncschuster** [comment]
Payment has been issued! Thanks for your contributions!
